shared/controlflow/codeql/controlflow/ControlFlow.qll

```diff
@@ -0,0 +1,894 @@
+    BaseValue(GuardValue gv) { possibleValue(_, gv) } or
+    IntRange(int low, int high) { possibleClosedRange(_, low, high) }
+
+  private class GuardValueExt extends TGuardValueExt {
+    string toString() {
+      result = "Any" and this = AnyValue()
+      or
+      result = this.asBase().toString()
+      or
+      exists(int low, int high |
+        this = IntRange(low, high) and result = "[" + low + ", " + high + "]"
+      )
+    }
+
+    GuardValue asBase() { this = BaseValue(result) }
+  }
+
+  private class TGuardValueOrAny = AnyValue or BaseValue;
+
+  private class GuardValueOrAny extends GuardValueExt, TGuardValueOrAny { }
+
+  private GuardValueExt mkRange(int low, int high) {
+    result = IntRange(low, high)
+    or
+    low = high and
+    result.asBase().asIntValue() = low
+  }
+
+  private GuardValueExt intersectBase1(GuardValue gv1, GuardValue gv2) {
+    exists(SourceVariable var |
+      possibleValue(var, gv1) and
+      possibleValue(var, gv2)
+    |
+      smaller(gv1, gv2) and result.asBase() = gv1
+      or
+      exists(int low, int high |
+        gv1.isIntRange(low, false) and
+        gv2.isIntRange(high, true) and
+        result = mkRange(low, high)
+      )
+    )
+  }
+
+  private GuardValueExt intersectBase2(GuardValueExt v1, GuardValue v2) {
+    result = intersectBase1(v1.asBase(), v2)
+    or
+    result = intersectBase1(v2, v1.asBase())
+  }
+
+  bindingset[v1, v2]
+  pragma[inline_late]
+  private GuardValueExt intersectRange(GuardValueExt v1, GuardValue v2) {
+    exists(int low, int high | v1 = IntRange(low, high) |
+      exists(int bound, boolean upper | v2.isIntRange(bound, upper) |
+        upper = true and result = mkRange(low, high.minimum(bound))
+        or
+        upper = false and result = mkRange(low.maximum(bound), high)
+      )
+      or
+      exists(int k |
+        v2.asIntValue() = k and
+        result.asBase() = v2 and
+        low <= k and
+        k <= high
+      )
+      or
+      not v2.isIntRange(_, _) and not exists(v2.asIntValue()) and result = v1
+    )
+  }
+
+  bindingset[v1, v2]
+  pragma[inline_late]
+  private GuardValueExt intersect(GuardValueExt v1, GuardValue v2) {
+    v1 = AnyValue() and result.asBase() = v2
+    or
+    result = intersectBase2(v1, v2)
+    or
+    result = v1 and
+    v1 instanceof BaseValue and
+    not exists(intersectBase2(v1, v2))
+    or
+    result = intersectRange(v1, v2)
+  }
+
+  bindingset[v1, gv2]
+  private predicate disjointValuesExt(GuardValueExt v1, GuardValue gv2) {
+    disjointValues(v1.asBase(), gv2)
+    or
+    exists(int low, int high | v1 = IntRange(low, high) |
+      gv2.asIntValue() < low
+      or
+      high < gv2.asIntValue()
+      or
+      exists(int bound, boolean upper | gv2.isIntRange(bound, upper) |
+        upper = true and bound < low
+        or
+        upper = false and high < bound
+      )
+    )
+  }
+
+  /** An input configuration for control flow reachability. */
+  signature module ConfigSig {
```

Comment:
Very nice that the predicates are named similar to those of dataflow configurations.